packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/ScmMethodProviderCollection.cs

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using Microsoft.Generator.CSharp.Expressions;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Generator.CSharp.Input;
+using Microsoft.Generator.CSharp.Providers;
+using Microsoft.Generator.CSharp.Statements;
+using static Microsoft.Generator.CSharp.Snippets.Snippet;
+using Microsoft.Generator.CSharp.Snippets;
+
+namespace Microsoft.Generator.CSharp.ClientModel.Providers
+{
+    internal class ScmMethodProviderCollection : MethodProviderCollection
+    {
+        public ScmMethodProviderCollection(InputOperation operation, TypeProvider enclosingType)
+            : base(operation, enclosingType)
+        {
+        }
+
+        protected override IReadOnlyList<MethodProvider> BuildMethods()
+        {
+            return
+            [
+                // TO-DO: Add Protocol and Convenience methods https://github.com/Azure/autorest.csharp/issues/4585, https://github.com/Azure/autorest.csharp/issues/4586
+                BuildCreateMessageMethod(_operation, _enclosingType),
+                BuildProtocolMethod(_operation, _enclosingType, false),
+                BuildProtocolMethod(_operation, _enclosingType, true)
+            ];
+        }
+        private static MethodProvider BuildProtocolMethod(InputOperation operation, TypeProvider enclosingType, bool isAsync)
+        {
+            List<ParameterProvider> methodParameters = new();
+            foreach (InputParameter inputParam in operation.Parameters)
+            {
+                if (inputParam.Kind != InputOperationParameterKind.Method)
+                    continue;
+                methodParameters.Add(ClientModelPlugin.Instance.TypeFactory.CreateCSharpParam(inputParam));
+            }
+
+            var methodModifier = MethodSignatureModifiers.Public | MethodSignatureModifiers.Virtual;
+            if (isAsync)
+            {
+                methodModifier |= MethodSignatureModifiers.Async;
+            }
+            var opName = operation.Name.ToCleanName();
+            var methodSignature = new MethodSignature(
+                isAsync ? opName + "Async" : opName,
+                FormattableStringHelpers.FromString(operation.Description),
+                methodModifier,
+                GetResponseType(operation.Responses, isAsync),
+                null,
+                Parameters: [.. methodParameters, KnownParameters.CancellationTokenParameter]);
```

Comment:
We take RequestOptions parameter in protocol methods right?

Reply:
this is just a wip and will be fixed before the issue closes but allows others to work off of this refactor